swupd/delta.go

```diff
@@ -0,0 +1,164 @@
+package swupd
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"sync"
+
+	"github.com/clearlinux/mixer-tools/helpers"
+	"github.com/pkg/errors"
+)
+
+// CreateDeltas creates all delta files between the previous and current
+// version of the supplied manifest, returning a list of files that
+// delta creation failed for.
+func CreateDeltas(manifest *Manifest, statedir string, from, to uint32) ([]*File, error) {
+	var c config
+	c, err := getConfig(statedir)
+	if err != nil {
+		return nil, err
+	}
+
+	var oldManifest *Manifest
+	var newManifest *Manifest
+
+	if oldManifest, err = ParseManifestFile(filepath.Join(c.outputDir, string(from), manifest.Name)); err != nil {
+		return nil, err
+	}
+	if newManifest, err = ParseManifestFile(filepath.Join(c.outputDir, string(to), manifest.Name)); err != nil {
+		return nil, err
+	}
+
+	_, _, _ = newManifest.linkPeersAndChange(oldManifest, 0)
+
+	deltas, err := consolidateDeltaFiles(newManifest, c)
+	if err != nil {
+		return nil, errors.Wrapf(err, "Failed to create deltas list %s", manifest.Name)
+	}
+
+	var failed []*File
+	var errchan = make(chan *File)
+	var wg sync.WaitGroup
+	for _, file := range deltas {
+		wg.Add(1)
+		go func(file *File, statedir string) {
+			defer wg.Done()
+			err := createDelta(file, statedir, file.DeltaPeer.Version, file.DeltaPeer.Hash)
+			if err != nil {
+				errchan <- file
+			}
+		}(file, statedir)
+	}
+
+	go func() {
+		wg.Wait()
+		close(errchan)
+	}()
+
+	for f := range errchan {
+		failed = append(failed, f)
+	}
+
+	return failed, nil
+}
+
+func consolidateDeltaFiles(manifest *Manifest, c config) ([]*File, error) {
+	var files []*File
+
+	if manifest == nil {
+		return nil, nil
+	}
+
+	for _, file := range manifest.Files {
+		if file.Version <= manifest.Header.Version ||
+			file.DeltaPeer == nil ||
+			file.Type != typeFile ||
+			file.DeltaPeer.Type != typeFile {
+			continue
+		}
+
+		deltadir := filepath.Join(c.outputDir, string(file.Version), "delta")
+		fromToString := fmt.Sprintf("%d-%d-%s-%s", file.DeltaPeer.Version, file.Version, file.DeltaPeer.Hash, file.Hash)
+
+		deltaFile := filepath.Join(deltadir, fromToString)
+
+		// Only add to list if the delta file does not already exist on the system
+		if _, err := os.Stat(deltaFile); os.IsNotExist(err) {
+			files = append(files, file)
+		}
+
+	}
+
+	return files, nil
+}
+
+func createDelta(file *File, statedir string, fromVersion uint32, fromHash Hashval) error {
+	var c config
+	c, err := getConfig(statedir)
```

Comment:
Oops, here is where I meant to comment we read config a lot

Reply:
Ohhh, yes this would read config a ton.

Reply:
I am also thinking that perhaps getConfig should return (*config, error) to reduce the amount of things that need to be copied around?

Reply:
Ah, should have scrolled more, I agree.

Reply:
This still needs to be fixed.